src/Numerics/Mesh/Metrics.jl

```diff
@@ -98,127 +100,298 @@ function creategrid!(x1, x2, x3, e2c, ξ1, ξ2, ξ3)
+- vgeo::VolumeGeometry, a struct containing the volumetric geometric factors
+- D::DAT2, 1-D derivative operator on the device in the first dimension
+
+Setup the variable needed for geometric factors data in vgeo.
+"""
+function setup_geom_factors_data!(
+    vgeo::VolumeGeometry{NTuple{3,Int}, <:AbstractArray},
+    D
+)
+    Nq = size(D, 1)
+    nelem = div(length(vgeo.ωJ), Nq)
+
+    vgeo.x1 = reshape(vgeo.x1, (Nq, nelem))
+    vgeo.ωJ = reshape(vgeo.ωJ, (Nq, nelem))
+    vgeo.JcV = reshape(vgeo.JcV, (Nq, nelem))
+    vgeo.ξ1x1 = reshape(vgeo.ξ1x1, (Nq, nelem))
+
+end
+
+"""
+    setup_geom_factors_data!(vgeo, D1, D2)
+
+Input arguments:
+- vgeo::VolumeGeometry, a struct containing the volumetric geometric factors
+- D1::DAT2, 1-D derivative operator on the device in the first dimension
+- D2::DAT2, 1-D derivative operator on the device in the second dimension
+
+Setup the variable needed for geometric factors data in vgeo.
+"""
+function setup_geom_factors_data!(
+    vgeo::VolumeGeometry{NTuple{3,Int}, <:AbstractArray},
+    D1,
+    D2
+)
+    Nq = (size(D1, 1), size(D2, 1))
+    nelem = div(length(vgeo.ωJ), prod(Nq))
+
+    vgeo.x1 = reshape(vgeo.x1, (Nq..., nelem))
+    vgeo.x2 = reshape(vgeo.x2, (Nq..., nelem))
+    vgeo.ωJ = reshape(vgeo.ωJ, (Nq..., nelem))
+    vgeo.JcV = reshape(vgeo.JcV, (Nq..., nelem))
+    vgeo.ξ1x1 = reshape(vgeo.ξ1x1, (Nq..., nelem))
+    vgeo.ξ2x1 = reshape(vgeo.ξ2x1, (Nq..., nelem))
+    vgeo.ξ1x2 = reshape(vgeo.ξ1x2, (Nq..., nelem))
+    vgeo.ξ2x2 = reshape(vgeo.ξ2x2, (Nq..., nelem))
+
+end
+
+"""
+    setup_geom_factors_data!(vgeo, D1, D2, D3)
+
+Input arguments:
+- vgeo::VolumeGeometry, a struct containing the volumetric geometric factors
+- D1::DAT2, 1-D derivative operator on the device in the first dimension
+- D2::DAT2, 1-D derivative operator on the device in the second dimension
+- D3::DAT2, 1-D derivative operator on the device in the third dimension
+
+Setup the variable needed for geometric factors data in vgeo.
+"""
+function setup_geom_factors_data!(
+    vgeo::VolumeGeometry{NTuple{3,Int}, <:AbstractArray},
+    D1,
+    D2,
+    D3,
+)
+    Nq = (size(D1, 1), size(D2, 1), size(D3, 1))
+    nelem = div(length(vgeo.ωJ), prod(Nq))
+
+    vgeo.x1 = reshape(vgeo.x1, (Nq..., nelem))
+    vgeo.x2 = reshape(vgeo.x2, (Nq..., nelem))
+    vgeo.x3 = reshape(vgeo.x3, (Nq..., nelem))
+    vgeo.ωJ = reshape(vgeo.ωJ, (Nq..., nelem))
+    vgeo.JcV = reshape(vgeo.JcV, (Nq..., nelem))
+    vgeo.ξ1x1 = reshape(vgeo.ξ1x1, (Nq..., nelem))
+    vgeo.ξ2x1 = reshape(vgeo.ξ2x1, (Nq..., nelem))
+    vgeo.ξ3x1 = reshape(vgeo.ξ3x1, (Nq..., nelem))
+    vgeo.ξ1x2 = reshape(vgeo.ξ1x2, (Nq..., nelem))
+    vgeo.ξ2x2 = reshape(vgeo.ξ2x2, (Nq..., nelem))
+    vgeo.ξ3x2 = reshape(vgeo.ξ3x2, (Nq..., nelem))
+    vgeo.ξ1x3 = reshape(vgeo.ξ1x3, (Nq..., nelem))
+    vgeo.ξ2x3 = reshape(vgeo.ξ2x3, (Nq..., nelem))
+    vgeo.ξ3x3 = reshape(vgeo.ξ3x3, (Nq..., nelem))
+
+end
+
+"""
+    compute_dxdxi_jacobian!(vgeo, D1, D2)
```

Comment:
might as well use unicode in the function name.

Reply:
If there a standard name in the finite element literature, that would be ideal.

Reply:
Might be better to make the function signature `compute_dxdxi_jacobian!(vgeo, (D1, D2))`, as then they're consistent.

Reply:
In commit [964adc8](https://github.com/CliMA/ClimateMachine.jl/pull/2162/commits/964adc8885a0172c740eaee49c01a209ff5b40a1) I decided to rename this function `compute_reference_to_physical_coord_jacobian` so that it doesn't need to use unicode and it is clear which coordinate transformation we are computing the Jacobian matrix of. Since soon we might have a longer chain rule of coordinate transformations, I thought it was better to be explicit in the name. Let me know what you think